enclave/core/seal.c

```diff
@@ -0,0 +1,192 @@
+        return OE_UNEXPECTED;
+
+    result = OE_NOT_FOUND;
+    for (i = 0; result != OE_OK && i < _num_plugins; ++i)
+        if (memcmp(&_plugins[i]->id, plugin_id, sizeof(*plugin_id)) == 0)
+        {
+            _plugins[i] = _plugins[--_num_plugins];
+            result = OE_OK;
+        }
+
+    oe_mutex_unlock(&_plugins_lock);
+    return result;
+}
+
+static const oe_seal_plugin_definition_t* _find_plugin(
+    const oe_uuid_t* plugin_id)
+{
+    int i;
+
+    if (plugin_id == NULL && _num_plugins > 0)
+        return _plugins[0];
+
+    for (i = 0; i < _num_plugins; ++i)
+        if (memcmp(&_plugins[i]->id, plugin_id, sizeof(*plugin_id)))
+            return _plugins[i];
+    return NULL;
+}
+
+oe_result_t oe_seal(
+    const oe_uuid_t* plugin_id,
+    const oe_seal_setting_t* settings,
+    size_t settings_count,
+    const uint8_t* plaintext,
+    size_t plaintext_size,
+    const uint8_t* additional_data,
+    size_t additional_data_size,
+    uint8_t** blob,
+    size_t* blob_size)
+{
+    oe_result_t result;
+    const oe_seal_plugin_definition_t* plugin;
+
+    if (blob == NULL || blob_size == NULL)
+        return OE_INVALID_PARAMETER;
+
+    *blob = NULL;
+    *blob_size = 0;
+
+    if ((settings == NULL) != (settings_count == 0) ||
+        (plaintext == NULL) != (plaintext_size == 0) ||
+        (additional_data == NULL) != (additional_data_size == 0))
+        return OE_INVALID_PARAMETER;
+
+    if (oe_mutex_lock(&_plugins_lock) != OE_OK)
+        return OE_UNEXPECTED;
+
+    plugin = _find_plugin(plugin_id);
+    if (plugin == NULL)
+        OE_RAISE(OE_NOT_FOUND);
+
+    result = plugin->seal(
+        settings,
+        settings_count,
+        plaintext,
+        plaintext_size,
+        additional_data,
+        additional_data_size,
+        blob,
+        blob_size);
+
+done:
+    oe_mutex_unlock(&_plugins_lock);
+    return result;
+}
+
+oe_result_t oe_unseal(
+    const uint8_t* blob,
+    size_t blob_size,
+    const uint8_t* additional_data,
+    size_t additional_data_size,
+    uint8_t** plaintext,
+    size_t* plaintext_size)
+{
+    oe_result_t result;
+    uint8_t* decrypted = NULL;
+    size_t decrypted_size;
+    int i;
+
+    if (blob == NULL || blob_size == 0 ||
+        (additional_data == NULL) != (additional_data_size == 0))
+        return OE_INVALID_PARAMETER;
+
+    if (plaintext == NULL)
+        plaintext = &decrypted;
+    if (plaintext_size == NULL)
+        plaintext_size = &decrypted_size;
+
+    *plaintext = NULL;
+    *plaintext_size = 0;
+
+    if (oe_mutex_lock(&_plugins_lock) != OE_OK)
+        return OE_UNEXPECTED;
+
+    for (i = 0; i < _num_plugins; ++i)
+    {
+        result = _plugins[i]->unseal(
+            blob,
+            blob_size,
+            additional_data,
+            additional_data_size,
+            plaintext,
+            plaintext_size);
+        if (result == OE_OK)
+            goto done;
+    }
+
+    result = OE_FAILURE;
```

Comment:
I would suggest the more descriptive `OE_UNSUPPORTED` in this case, which would also be applicable for a corrupted seal header. 

It's also an option to add new error codes to result.h if you think the cases are common enough to warrant better granularity errors.

Reply:
`OE_FAILURE` means failed for an unspecified reason, while `OE_UNSUPPORTED` means a feature requested was not available. Given the cause could be due to a tampered blob or a missing plug-in, I do think `OE_FAILURE` is better.